clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -348,10 +347,15 @@ private TransactionalRequestResult beginCompletingTransaction(TransactionResult
             isTransactionV2Enabled
         );
 
+        // Maybe update the transaction version here before we enqueue the EndTxn request so there are no races with
```

Comment:
We should probably clarify that this function may set the `clientSideEpochBumpRequired` on upgrade and that's the reason we do it here, before we call logic that checks `clientSideEpochBumpRequired`.  But not before we build the EndTxnRequest, because we need to complete the transaction with the version it started.